Prevent Aliased Matchers from Overriding Expected Data

lnestor opened the pull request:
## Previous Behavior

The aliased matcher implementation of a description currently replaces all instance of the old matcher's name in the new matcher's name. If it happens that the expected values include the old matcher's name, these also get overwritten. In the examples below, the string `"include"` is replaced in both the outputs by the new matcher's name.

```ruby
it 'overwrites data for a_string_including' do
  my_string = "a string with 'include'"
  expect("some random string").to match(a_string_including(my_string))
end

# => expected "some random string" to match 
#       (a string including "a string with 'a string including'")
```

```ruby
it 'overwrites data for a_hash_including' do
  my_hash = { my_string: "a string with 'include'" }
  expect({}).to match(a_hash_including(my_hash))
end

# => expected {} to match 
#       (a hash including {:my_string => "a string with 'a hash including'"})
```

## New Behavior

This change makes it so it only replaces the first instance of the original name so none of the expected data is also overwritten. The two above error outputs would be the following:

`expected {} to match (a hash including {:my_string => "a string with 'include'"})`
`expected "some random string" to match (a string including "a string with 'include'")`

I saw this in https://github.com/rspec/rspec-rails/issues/1835.

Commit:
Prevent aliased matchers from overriding data

The aliased matcher implementation of a description currently
replaces all instance of the old matcher's name with the new
matcher's name. However, in cases where the expected value
has the same string as the old matcher's name, it would also
get replaced. This change makes it so only the first instance
of the old matcher's name is replaced in the description.

JonRowe commented on lib/rspec/matchers/dsl.rb:
```diff
@@ -29,7 +29,7 @@ module DSL
       def alias_matcher(new_name, old_name, options={}, &description_override)
         description_override ||= lambda do |old_desc|
-          old_desc.gsub(EnglishPhrasing.split_words(old_name), EnglishPhrasing.split_words(new_name))
+          old_desc.sub(EnglishPhrasing.split_words(old_name), EnglishPhrasing.split_words(new_name))
```
I think this needs to only match the start of the string, this might still accidentally override terms if the original description is not generated.

lnestor replied:
Thanks for the feedback! Can you clarify this a bit? `#sub` matches the first occurrence, so are you saying something like matching the first n characters of the string?

Also, what do you mean by if the original description is not generated? Since `BaseMatcher` implements description, I would think that they all would have a description. Unless a class implements its own description which does not contain the name of the matcher.

JonRowe replied:
> Can you clarify this a bit? #sub matches the first occurrence, so are you saying something like matching the first n characters of the string?

`"hello_old_name".sub("old_name","new_name")` would produce `"hello_new_name"` but we don't want that. We only want to replace `old_name` as it was generated initially. 

> Also, what do you mean by if the original description is not generated? Since BaseMatcher implements description, I would think that they all would have a description. 

Similarly it is possible for matchers to implement `description` themselves and it might mean they don't contain the old_name at all, in this case we shouldn't be replacing the first instance blindly as it might not be in the description but in the expected.

JonRowe commented on spec/rspec/matchers/aliased_matcher_spec.rb:
```diff
@@ -79,6 +79,19 @@ def description
+
+      it 'does not override data in the description with the same name as the matcher' do
+        matcher = my_repeating_override
+        expect(matcher.description).to eq("my repeating override my repeating base matcher")
```
It'd prefer a practical example showing the problem using `fail_with`, even the one from rspec/rspec-rails#1835 would suffice.

JonRowe reviewed (changes requested):
Thanks for tackling this, a bit of feedback above!

Commit:
Give option to override aliased matcher name no matter the position

Commit:
Remove lambda default parameter to maintain ruby compatability

Commit:
Split aliased matcher description into new class

A flag was added when generating an aliased matcher's
description to better control how the old matcher's name
is replaced. However, this broke the existing API for
overriding a matcher's new description. This class allows
to maintain backwards compatability gracefully.

Commit:
Remove keyword arguments for backwards compatibility

JonRowe commented on lib/rspec/matchers/aliased_matcher.rb:
```diff
@@ -105,12 +105,53 @@ def failure_message_when_negated
+          old_desc
+        end
+      end
+    end
```
This is a lot of added complication, and still doesn't remove the global sub in favour of a "start of string only" replace.

pirj commented:
Do you have any questions on how to proceed with this @lnestor ?

pirj commented:
@lnestor Are you interested in wrapping this up?

JonRowe commented:
Migrated to the monorepo as rspec/rspec#123
